k8s/src/main/scala/io/buoyant/k8s/Watchable.scala

```diff
@@ -139,71 +121,106 @@ private[k8s] abstract class Watchable[O <: KubeObject: TypeReference, W <: Watch
-              // stream, but this is necessary to handle the K8s bug where
-              // erorrs have the incorrect status code.
-              .flatMap {
-                // special case to handle Kubernetes bug where "too old
-                // resource version" errors are returned with status code 200
-                // rather than status code 410.
-                // see https://github.com/kubernetes/kubernetes/issues/35068
-                // for details.
-                case e: Watch.Error[O] if e.status.code.contains(410) =>
-                  log.debug(
-                    "k8s returned 'too old resource version' error with " +
-                      "incorrect HTTP status code, restarting watch"
-                  )
-                  _resourceVersionTooOld()
-                case event => AsyncStream.fromOption(Some(event))
-              }
-            watchStream ++ // if the stream ends (k8s will kill connections after ~30m), restart it)
-              AsyncStream.fromFuture {
-                // It's safe to call lastOption here, because we're after the `++` operator, so
-                // this code will not be executed until `watchStream` has completed.
-                val currentVersion = watchStream.lastOption.map(_.flatMap(_.resourceVersion))
-                currentVersion.flatMap {
-                  case Some(version) => Future.value(_watch(Some(version)))
-                  case None =>
-                    // In this case, we want to try loading the initial information instead before watching again.
-                    restartWatches(labelSelector, fieldSelector).map {
-                      case (ws, ver) => AsyncStream.fromSeq(ws) ++ _watch(ver)
-                    }
-                }
-              }
-              .flatten
+
+            _processEventStream(
+              () => Json.readStream[W](rsp.reader, Api.BufSize),
+              resourceVersion
+            )
 
           case http.Status.Gone =>
             _resourceVersionTooOld()
 
           case status =>
             close.set(Closable.nop)
             log.debug("k8s failed to watch resource %s: %d %s", path, status.code, status.reason)
-            val f = Future.exception(Api.UnexpectedResponse(rsp))
-            AsyncStream.fromFuture(f)
+            val exc = Api.UnexpectedResponse(rsp)
+            state(Activity.Failed(exc))
+            Future.exception(exc)
         }
       }
-        // Ignore any cases where we receive a resource version lower
-        // than the last received resource version.
-        .increasingOnly
+    }
+
+    /**
+     * Recursively consume an AsyncStream of watch events from an established connection,
+     * updating [[state]] when appropriate.  Restart the watch on various error conditions.
+     */
+    def _processEventStream(
+      stream: () => AsyncStream[W],
+      resourceVersion: Option[String],
+      previousEvent: Option[W] = None
```

Comment:
consider calling this "largestVersion" or something since it may not actually be the previous event

Reply:
updated to 'largestVersion' and 'largestEvent'.  

Aside: a good cleanup to do afterwards may be to get rid of the event argument here.  I think the resource version comparison from ResourceVersionOrdering can be factored out and used directly with the resource version here, avoiding the need to keep a whole event around.  (I think this may be the only use case of that Ordering type class ...  I didn't do the change wanting to limit the diff size.)